### PR TITLE
Fix #4434:Dashboard backend admin service URL is not configurable

### LINKED ISSUE #4434: Dashboard backend admin service URL is not configurable
If we are setting up a proxy for the identity server and set proxy host and port as mentioned in [doc](https://docs.wso2.com/display/IS530/Setting+Up+Deployment+Pattern+1#SettingUpDeploymentPattern1-Changinghostnamesandports), then all the admin service calls from the dashboard and portal jaggery apps are using `https://proxy_host:proxyport/services` URL.

There are some cases, we need to avoid these admin service calls going through proxy and hit the backend direclty.

According to [doc](https://docs.wso2.com/display/IS530/Setting+Up+Deployment+Pattern+1#SettingUpDeploymentPattern1-Settingupthedashboard), it seems that URL is configurable using `site.json`. But according to the [code](https://github.com/wso2/product-is/blob/master/modules/jaggery-apps/user-dashboard/dashboard/util/utility.jag), the `proxyHost` and `proxyPort` set in `site.json` is discarded in the `setServerURL()`.

### PR DESCRIPTION
**Pull request by IsurangaPerera**
Fix wso2/product-is#4434

### CODE REVIEW

File: modules/jaggery-apps/user-dashboard/dashboard/util/utility.jag [hunk 1]
```diff
@@ -259,133 +259,61 @@ String.prototype.replaceAll = function(stringToFind,stringToReplace){
...
-            application.put(SERVER_CORE_URL, serverCoreUrl);
-        }
-        else{
-            serverCoreUrl = identityServerURL.getServerURL("", true, true);
-            application.put(SERVER_CORE_URL, serverCoreUrl);
-        }
-    }
 }
+
 %>
```
Comment (isharailanga) [inline modules/jaggery-apps/user-dashboard/dashboard/util/utility.jag]: Shall we add a new line in the EOF

File: modules/jaggery-apps/portal/conf/site.json [hunk 2]
```diff
@@ -1,8 +1,8 @@
 {
     "proxy" : {
-        "proxyHost" : "",
-        "proxyHTTPSPort" : "",
-        "proxyContextPath" : ""
+        "serverOrigin" : "https://localhost:9443",
+        "proxyContextPath" : "",
+        "servicePath" : "/services"
```
Comment (malithie) [inline modules/jaggery-apps/portal/conf/site.json]: IMO, having serverOrigin is enough. For me it seems more intuitive if we rename it as 'serviceOrigin'
Other configs were not used anyway as at now and servicesPath may differ when it comes to REST APIs.

File: modules/jaggery-apps/portal/util/utility.jag [hunk 3]
```diff
@@ -95,24 +43,20 @@ function setServerURL() {
...
-
-        //basic server core url : https://<HOST_NAME>:<PORT>/<PROXY_CONTEXT_PATH>
-        var serverCoreUrl = identityServerURL.getServerURL("", true, false);
+        if ( serverOrigin == null || serverOrigin === '' || serverOrigin === 'null' || serverOrigin.length <= 0 ){
+            serverCoreUrl = identityServerURL.getServerURL("", true, false);
+            serverUrl = identityServerURL.getServerURL("services", true, true);
+        } else {
+            serverCoreUrl = serverOrigin + proxyContextPath;
+            serverUrl = serverCoreUrl + "/services";
+        }
```
Comment (malithie) [inline modules/jaggery-apps/portal/util/utility.jag]: IMO, unless severOrigin which you have added not configured you can build from the IdentityUtil. That handles proxy context paths anyway.

File: modules/jaggery-apps/portal/util/utility.jag [hunk 4]
```diff
@@ -95,24 +43,20 @@ function setServerURL() {
...
         }
 
-        //create server service url
-        serverUrl = identityServerURL.getServerURL("", true, true) + "/services";
-
-        //basic server core url : https://<HOST_NAME>:<PORT>/<PROXY_CONTEXT_PATH>
-        var serverCoreUrl = identityServerURL.getServerURL("", true, false);
+        if ( serverOrigin == null || serverOrigin === '' || serverOrigin === 'null' || serverOrigin.length <= 0 ){
+            serverCoreUrl = identityServerURL.getServerURL("", true, false);
+            serverUrl = identityServerURL.getServerURL("services", true, true);
```
Comment (malithie) [inline modules/jaggery-apps/portal/util/utility.jag]: Let's rename this as 'adminServiceUrl' or 'serviceUrl'

File: modules/jaggery-apps/user-dashboard/dashboard/apis/ues.jag [hunk 5]
```diff
@@ -2,9 +2,8 @@
 response.addHeader("Access-Control-Allow-Origin", "*");
 response.addHeader("Content-Type", "application/json");
 
-var util = require("/util/utility.jag");
-util.setServerURL("false");
-var serverHttpsUrl = application.get(SERVER_CORE_URL);
+var identityServerURL = Packages.org.wso2.carbon.identity.core.util.IdentityUtil;
+var serverHttpsUrl = identityServerURL.getServerURL("", true, false);
```
Comment (malithie) [inline modules/jaggery-apps/user-dashboard/dashboard/apis/ues.jag]: For what is this variable being used for

File: modules/jaggery-apps/user-dashboard/dashboard/conf/site.json [hunk 6]
```diff
@@ -1,7 +1,6 @@
 {
-    "proxy" : {
-        "proxyHost" : "",
-        "proxyHTTPSPort" : "",
+    "proxy": {
+        "serverOrigin" : "https://localhost:9443",
```
Comment (malithie) [inline modules/jaggery-apps/user-dashboard/dashboard/conf/site.json]: Please refer the comment I have placed above for the same config

File: modules/jaggery-apps/user-dashboard/dashboard/index.jag [hunk 7]
```diff
@@ -3,27 +3,22 @@
...
+            } else {
+                response.sendRedirect("login.jag");
+            }
         }
 
-        var httpsCoreUrl = application.get(SERVER_CORE_URL);
-        if (httpsCoreUrl === 'null' || httpsCoreUrl == null || httpsCoreUrl.length <= 0) {
-        httpsCoreUrl = '';
-        }
+        var httpsCoreUrl = identityServerURL.getServerURL("", true, true);
```
Comment (malithie) [inline modules/jaggery-apps/user-dashboard/dashboard/index.jag]: Why read this back again from the util class. Shouldn't the SERVER_CORE_URL being populated should be used here

File: modules/jaggery-apps/user-dashboard/dashboard/util/utility.jag [hunk 8]
```diff
@@ -259,133 +259,61 @@ String.prototype.replaceAll = function(stringToFind,stringToReplace){
...
+        if ( serverOrigin == null || serverOrigin === '' || serverOrigin === 'null' || serverOrigin.length <= 0 ){
+            serverUrl = identityServerURL.getServerURL("services", true, true);
+            if (isWebContextRootAvailable === 'false') {
+                serverCoreUrl = identityServerURL.getServerURL("", true, false);
+            } else {
+                serverCoreUrl = identityServerURL.getServerURL("", true, true);
+            }
+        } else {
+            serverCoreUrl = serverOrigin + proxyContextPath;
+            serverUrl = serverCoreUrl + "/services";
```
Comment (malithie) [inline modules/jaggery-apps/user-dashboard/dashboard/util/utility.jag]: Both apps should include the same logic

File: modules/jaggery-apps/user-dashboard/dashboard/apis/ues.jag (hunk 5 above)
Comment (IsurangaPerera) [reply modules/jaggery-apps/user-dashboard/dashboard/apis/ues.jag]: it is appended to each gadget so that it can be fetched from dashboard

**Comment by malithie** [comment]
@IsurangaPerera ,
Please mention the approach you followed in fixing this issue

File: modules/jaggery-apps/portal/util/utility.jag [hunk 9]
```diff
@@ -95,24 +35,19 @@ function setServerURL() {
...
-        serverUrl = identityServerURL.getServerURL("", true, true) + "/services";
-
-        //basic server core url : https://<HOST_NAME>:<PORT>/<PROXY_CONTEXT_PATH>
-        var serverCoreUrl = identityServerURL.getServerURL("", true, false);
+        if ( serviceOrigin == null || serviceOrigin === '' || serviceOrigin === 'null' || serviceOrigin.length <= 0 ){
+            serverCoreUrl = identityServerURL.getServerURL("", true, false);
+            serverUrl = identityServerURL.getServerURL("services", true, true);
+        } else {
+            serverCoreUrl = serviceOrigin + proxyContextPath;
+            serverUrl = serverCoreUrl + "/services";
```
Comment (malithie) [inline modules/jaggery-apps/portal/util/utility.jag]: IMO, we should not append proxyContextPath for this configured serviceOrigin URL. That's expected to be configured with the exact path. 
This respective origin could point to the internal server as well, which conflicts when you append proxycontextPath derived from proxyhost and webcontext root

File: modules/jaggery-apps/user-dashboard/dashboard/controllers/login-logout/SAML2SSOAuthenticationClient.jag [hunk 10]
```diff
@@ -6,7 +6,7 @@ var identityUtil = Packages.org.wso2.carbon.identity.core.util.IdentityUtil;
 var log = new Log();
 var serverUrl = util.getServerUrl();
 var serverUrl2 = serverUrl + "/SAML2SSOAuthenticationService";
-var identityServerUrl = identityUtil.getServerURL(getServicePath() + "/SAML2SSOAuthenticationService", true, true);
+var identityServerUrl = identityUtil.getServerURL("services" + "/SAML2SSOAuthenticationService", true, true);
```
Comment (malithie) [inline modules/jaggery-apps/user-dashboard/dashboard/controllers/login-logout/SAML2SSOAuthenticationClient.jag]: This should be from the 'SEVER_URL' built above

File: modules/jaggery-apps/user-dashboard/dashboard/util/utility.jag (hunk 8 above)
Comment (malithie) [reply modules/jaggery-apps/user-dashboard/dashboard/util/utility.jag]: Refer comment above

File: modules/jaggery-apps/portal/conf/site.json [hunk 11]
```diff
@@ -1,8 +1,6 @@
 {
     "proxy" : {
-        "proxyHost" : "",
-        "proxyHTTPSPort" : "",
-        "proxyContextPath" : ""
+        "serviceOrigin" : "https://localhost:9443"
```
Comment (malithie) [inline modules/jaggery-apps/portal/conf/site.json]: Did we check on using placeholder for the port

**Comment by malithie** [comment]
@IsurangaPerera , Did we check the behavior of webcontextroot that can be configured in carbon.xml

File: modules/jaggery-apps/portal/conf/site.json [hunk 12]
```diff
@@ -1,10 +1,8 @@
 {
     "proxy" : {
-        "proxyHost" : "",
-        "proxyHTTPSPort" : "",
-        "proxyContextPath" : ""
+        "serviceOrigin" : "${carbon.protocol}://${carbon.host}:${carbon.management.port}"
```
Comment (malithie) [inline modules/jaggery-apps/portal/conf/site.json]: The 'serviceOrigin' does not match with 'proxy' related configs. Shall we introduce a new section like 'service' or 'server'.

**Pull request merged**